### PR TITLE
support bitbucket load workspace,projects,repositories,members

### PR DESCRIPTION
**Pull request by sagarnavale**

### CODE REVIEW

File: cartography/data/jobs/cleanup/bitbucket_workspace_member_cleanup.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "statements": [
+        {
+            "query": "MATCH (:BitbucketMember)<-[r:MEMBER]-(:BitbucketWorkspace{uuid: $WORKSPACE_UUID})<-[:OWNER]-(:CloudanixWorkspace{id: $WORKSPACE_ID}) WHERE r.lastupdated <> $UPDATE_TAG WITH r LIMIT $LIMIT_SIZE DELETE (r) return COUNT(*) as TotalCompleted",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/bitbucket_workspace_member_cleanup.json:4]: Clean up node first then relationship

File: cartography/data/jobs/cleanup/bitbucket_workspace_project_cleanup.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+    "statements": [
+        {
+            "query": "MATCH (:BitbucketProjects)<-[r:PROJETS]-(:BitbucketWorkspace{uuid: $WORKSPACE_UUID})<-[:OWNER]-(:CloudanixWorkspace{id: $WORKSPACE_ID}) WHERE r.lastupdated <> $UPDATE_TAG WITH r LIMIT $LIMIT_SIZE DELETE (r) return COUNT(*) as TotalCompleted",
```
Comment (ganeshjagdale9545) [inline cartography/data/jobs/cleanup/bitbucket_workspace_project_cleanup.json:4]: Same as above

File: cartography/cli.py [hunk 3]
```diff
@@ -853,4 +856,40 @@ def run_github(request):
...
+
+    default_sync = cartography.sync.build_bitbucket_sync()
+    config = Config(
+        request['neo4j']['uri'],
+        neo4j_user=request['neo4j']['user'],
+        neo4j_password=request['neo4j']['pwd'],
+        neo4j_max_connection_lifetime=request['neo4j']['connection_lifetime'],
+        params=request['params'],
+        bitbucket_refresh_token=get_access_token(request['client_id'],request['client_secret'],request['refresh_token']),
+        gcp_requested_syncs=request.get('services', None),
```
Comment (ganeshjagdale9545) [inline cartography/cli.py:874]: this is for gcp

File: cartography/cli.py [hunk 4]
```diff
@@ -853,4 +856,40 @@ def run_github(request):
...
+    )
+
+    if request['logging']['mode'] == "verbose":
+        config.verbose = True
+    elif request['logging']['mode'] == "quiet":
+        config.quiet = True
+
+    return CLI(default_sync, prog='cartography').process(config)
+
+def get_access_token( client_id: str, client_secret: str, refresh_token: str):
```
Comment (ganeshjagdale9545) [inline cartography/cli.py:884]: move this to the worker file

File: cartography/intel/bitbucket/__init__.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+def _sync_multiple_workspaces(
+    neo4j_session: neo4j.Session,
+    refresh_token: str,
+    workspaces:List[Dict],
+    common_job_parameters: Dict[str, Any],
+    ) ->bool:
+    
+    
+    for workspace in workspaces:
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/__init__.py:67]: add logic here to update `WORKSPACE_UUID` for each workspace in common_job_parameters

File: cartography/intel/bitbucket/__init__.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        _sync_one_workspace(neo4j_session,workspace.get('name'),refresh_token,common_job_parameters)
+
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/__init__.py:69]: also clean up for workspace

File: cartography/intel/bitbucket/__init__.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    del common_job_parameters['WORKSPACE_UUID']
+    return True
+
+def _sync_one_workspace(
+    neo4j_session: neo4j.Session,
+    workspace_name:str,
+    refresh_token:str,
+    common_job_parameters: Dict[str, Any],
+  ):
+    cartography.intel.bitbucket.projects.sync(
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/__init__.py:79]: can you use resources mapping logic like used in other clouds refer resources.py file

File: cartography/intel/bitbucket/members.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    url = f"https://api.bitbucket.org/2.0/workspaces/{workspace}/members"
+    return make_requests_url(url,access_token)
+
+
+
+    
+def load_memebers_data(neo4j_session: neo4j.Session,members_data:List[Dict],common_job_parameters:Dict):
+    ingest_workspace="""
+    UNWIND $membersData as member
+    MERGE (mem:BitbucketMember{name: member.user.display_name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/members.py:24]: use unique id to load resources if not present then create with standard format

File: cartography/intel/bitbucket/members.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    ON CREATE SET mem.firstseen = timestamp()
+
+    SET mem.slug = member.slug,
+    mem.type = member.user.type,
+    mem.account_id=member.user.account_id,
+    mem.uuid = member.user.uuid,
+    mem.lastupdated = $UpdateTag
+
+    WITH mem,member
+    MATCH (owner:BitbucketWorkspace{name:member.workspace.name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/members.py:34]: same here match with unique id

File: cartography/intel/bitbucket/members.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+@timeit
+def get_workspace_members(access_token:str,workspace:str):
+    url = f"https://api.bitbucket.org/2.0/workspaces/{workspace}/members"
+    return make_requests_url(url,access_token)
+
+
+
+    
+def load_memebers_data(neo4j_session: neo4j.Session,members_data:List[Dict],common_job_parameters:Dict):
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/members.py:21]: use transaction base approach to load resources 
https://github.com/Cloudanix/cartography/blob/54a7e1ea0f3cd0522fee67e32ad08bab883ae8c8/cartography/intel/azure/aks.py#L21

File: cartography/intel/bitbucket/members.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    SET o.lastupdated = $UpdateTag
+
+    """
+    neo4j_session.run(
+        ingest_workspace,
+        membersData=members_data,
+        UpdateTag=common_job_parameters['UPDATE_TAG'],
+    )
+    
+    cleanup(neo4j_session,common_job_parameters)
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/members.py:46]: Remove this outside the function

File: cartography/intel/bitbucket/workspace.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            name=workspace.get("name"),
+            created_on=workspace.get('created_on'),
+            slug=workspace.get('slug'),
+            type=workspace.get('type'),
+            uuid=workspace.get('uuid'),
+            is_private=workspace.get('is_private'),
+            UpdateTag=common_job_parameters['UPDATE_TAG'],
+            workspace_id=common_job_parameters['WORKSPACE_ID'],
+        )
+        cleanup(neo4j_session,workspace.get('uuid'),common_job_parameters)
```
Comment (ganeshjagdale9545) [inline cartography/intel/bitbucket/workspace.py:54]: Don’t clean up parent node first . first clean up all child node then delete parent

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
few changes needed

File: cartography/data/jobs/cleanup/bitbucket_workspace_member_cleanup.json (hunk 1 above)
Comment (sagarnavale) [reply cartography/data/jobs/cleanup/bitbucket_workspace_member_cleanup.json:4]: change first node clean then relationship

File: cartography/data/jobs/cleanup/bitbucket_workspace_project_cleanup.json (hunk 2 above)
Comment (sagarnavale) [reply cartography/data/jobs/cleanup/bitbucket_workspace_project_cleanup.json:4]: change first node clean then relationship

File: cartography/cli.py (hunk 3 above)
Comment (sagarnavale) [reply cartography/cli.py:874]: changed as bitbucket

File: cartography/intel/bitbucket/__init__.py (hunk 5 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/__init__.py:67]: updated new workspace  uuid

File: cartography/intel/bitbucket/__init__.py (hunk 6 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/__init__.py:69]: chnaged

File: cartography/intel/bitbucket/members.py (hunk 8 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/members.py:24]: uuid unique attribute  as id

File: cartography/intel/bitbucket/members.py (hunk 9 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/members.py:34]: uuid unique attribute as id

File: cartography/intel/bitbucket/members.py (hunk 10 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/members.py:21]: used transaction based approach

File: cartography/intel/bitbucket/workspace.py (hunk 12 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/workspace.py:54]: changes cleanup workspace after child node

File: cartography/intel/bitbucket/members.py (hunk 11 above)
Comment (sagarnavale) [reply cartography/intel/bitbucket/members.py:46]: moved to sync

**Review by ganeshjagdale9545 — APPROVED** [review]
Looks good

File: cartography/intel/azure/iam.py [hunk 13]
```diff
@@ -1,743 +1,927 @@
+import enum
+import json
 import logging
-import math
-from typing import Dict
-from typing import List
+import time
+from typing import Any, Dict, List, Tuple
 
+import boto3
```
Comment (ganeshjagdale9545) [inline cartography/intel/azure/iam.py:7]: can you check this file
you have moved aws specific code here

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
one changes needed

**Review by mpurusottamc — COMMENTED** [review]
looks okay. but, will merge after github changes are done.

**Review by mpurusottamc — APPROVED** [review]
merging for local testing and deployment

**Pull request merged**